earth_enterprise/src/fusion/autoingest/sysman/AssetHandleD.h

```diff
@@ -265,26 +174,27 @@ class MutableAssetHandleD_ : public virtual Base_ {
     if (dirtyMap.size()) {
       fprintf(stderr, "========== %d dirty %s ==========\n",
               dirtyMap.size(), header.c_str());
-      for (typename DirtyMap::const_iterator d = dirtyMap.begin();
-           d != dirtyMap.end(); ++d) {
+      std::vector<std::string> dirtyKeys = dirtyMap.keys();
+      for (auto& d : dirtyKeys) {
         fprintf(stderr, "%s -> %p\n",
-                d->first.c_str(), d->second.operator->());
+                d.c_str(), dirtyMap[d].operator->());
       }
     }
   }
 
 
   static bool SaveDirtyToDotNew(khFilesTransaction &savetrans,
                                 std::vector<std::string> *saveDirty) {
-    for (typename DirtyMap::const_iterator d = dirtyMap.begin();
-         d != dirtyMap.end(); ++d) {
+    DirtyMap tempDirty = std::move(dirtyMap);
+    std::vector<std::string> dirtyKeys = tempDirty.keys();
```

Comment:
I think there's still an issue here. Once you call std::move on dirtyMap it's no longer valid, so it needs to be reinitialized. I think that simply calling dirtyMap.keys() and then dirtyMap.clear() while holding a lock would do the trick. Similarly, if you just call dirtyMap.keys() and then remove the keys one by one so that you only remove the keys that you've handled would also work.